### PR TITLE
fix: reduce flakiness of LoadFast4Pwa

### LINKED ISSUE #2267: FastEnough Audit should be more specific to avoid accidental failures
Let's only look at first req for each origin. 

Also, ask them to file a bug if our throttling is on and it fails.

### PR DESCRIPTION
**Pull request by patrickhulce**
closes #2267 

This doesn't actually seem as critical now due to #2238, but confirmed this fixes the issue for https://shop.polymer-project.org/ from master before #2238

### CODE REVIEW

File: lighthouse-core/lib/url-shim.js [hunk 1]
```diff
@@ -74,6 +74,18 @@ URL.originsMatch = function originsMatch(urlA, urlB) {
 
 /**
  * @param {string} url
+ * @return {string|undefined}
```
Comment (brendankenny) [inline lighthouse-core/lib/url-shim.js]: nit, maybe return `null` in the failure case? A little more explicit

File: lighthouse-core/audits/load-fast-enough-for-pwa.js [hunk 2]
```diff
@@ -52,27 +53,35 @@ class LoadFastEnough4Pwa extends Audit {
...
-
-        return {
+        const latencyInfo = {
           url: record._url,
+          startTime: record._startTime,
+          origin: origin,
           latency: latency.toLocaleString(undefined, {maximumFractionDigits: 2})
         };
+
+        // Only examine the first request per origin to reduce noisiness
```
Comment (brendankenny) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: can we give a more specific justification for this?

File: lighthouse-core/audits/load-fast-enough-for-pwa.js [hunk 3]
```diff
@@ -52,27 +53,35 @@ class LoadFastEnough4Pwa extends Audit {
   static audit(artifacts) {
     const devtoolsLogs = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
     return artifacts.requestNetworkRecords(devtoolsLogs).then(networkRecords => {
-      const allRequestLatencies = networkRecords.map(record => {
-        // Ignore requests that don't have timing data or resources that have
-        // previously been requested and are coming from the cache.
+      const firstRequestLatenciesByOrigin = new Map();
+      networkRecords.forEach(record => {
+        // Ignore requests that don't have valid origin, timing data, or came from the cache.
```
Comment (ebidel) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: , or are not finished.

File: lighthouse-core/audits/load-fast-enough-for-pwa.js [hunk 4]
```diff
@@ -52,27 +53,35 @@ class LoadFastEnough4Pwa extends Audit {
...
         }
 
         // Use DevTools' definition of Waiting latency: https://github.com/ChromeDevTools/devtools-frontend/blob/66595b8a73a9c873ea7714205b828866630e9e82/front_end/network/RequestTimingView.js#L164
         const latency = record._timing.receiveHeadersEnd - record._timing.sendEnd;
-
-        return {
+        const latencyInfo = {
           url: record._url,
+          startTime: record._startTime,
+          origin: origin,
```
Comment (ebidel) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: nit: `origin,`

File: lighthouse-core/audits/load-fast-enough-for-pwa.js [hunk 5]
```diff
@@ -52,27 +53,35 @@ class LoadFastEnough4Pwa extends Audit {
...
+        if (!existing || existing.startTime > latencyInfo.startTime) {
+          firstRequestLatenciesByOrigin.set(origin, latencyInfo);
+        }
       });
 
+      const firstRequestLatencies = Array.from(firstRequestLatenciesByOrigin.values());
       const latency3gMin = Emulation.settings.TYPICAL_MOBILE_THROTTLING_METRICS.targetLatency - 10;
-      const areLatenciesAll3G = allRequestLatencies.every(val =>
-          val === undefined || val.latency > latency3gMin);
+      const areLatenciesAll3G = firstRequestLatencies.every(val => val.latency > latency3gMin);
```
Comment (ebidel) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: is  `val.latency` going to throw if you have undefines from 63?

File: lighthouse-core/audits/load-fast-enough-for-pwa.js (hunk 3 above)
Comment (patrickhulce) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: done

File: lighthouse-core/audits/load-fast-enough-for-pwa.js (hunk 4 above)
Comment (patrickhulce) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: done

File: lighthouse-core/audits/load-fast-enough-for-pwa.js (hunk 5 above)
Comment (patrickhulce) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: no it's a `forEach` now not a map, `firstRequestLatencies` are picked from the values of the map

File: lighthouse-core/lib/url-shim.js (hunk 1 above)
Comment (patrickhulce) [inline lighthouse-core/lib/url-shim.js]: done

File: lighthouse-core/audits/load-fast-enough-for-pwa.js (hunk 2 above)
Comment (patrickhulce) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: added a line about h2 push

File: lighthouse-core/audits/load-fast-enough-for-pwa.js (hunk 5 above)
Comment (ebidel) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: Ah yea. Good call.

File: lighthouse-core/lib/url-shim.js [hunk 6]
```diff
@@ -74,6 +74,18 @@ URL.originsMatch = function originsMatch(urlA, urlB) {
 
 /**
  * @param {string} url
+ * @return {?string}
+ */
+URL.getOrigin = function getOrigin(url) {
```
Comment (brendankenny) [inline lighthouse-core/lib/url-shim.js]: add a test for `getOrigin`? Maybe add whatever is in the network records that don't have valid origins as the bad test case

File: lighthouse-core/audits/load-fast-enough-for-pwa.js [hunk 7]
```diff
@@ -52,27 +53,36 @@ class LoadFastEnough4Pwa extends Audit {
...
           url: record._url,
+          startTime: record._startTime,
+          origin,
           latency: latency.toLocaleString(undefined, {maximumFractionDigits: 2})
         };
+
+        // Only examine the first request per origin to reduce noisiness from cases like H2 push
+        // where individual request latency may not apply.
+        const existing = firstRequestLatenciesByOrigin.get(origin);
+        if (!existing || existing.startTime > latencyInfo.startTime) {
```
Comment (paulirish) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: if (!existing || latencyInfo.startTime < existing.startTime) {

File: lighthouse-core/audits/load-fast-enough-for-pwa.js [hunk 8]
```diff
@@ -52,27 +53,36 @@ class LoadFastEnough4Pwa extends Audit {
...
+          firstRequestLatenciesByOrigin.set(origin, latencyInfo);
+        }
       });
 
+      const firstRequestLatencies = Array.from(firstRequestLatenciesByOrigin.values());
       const latency3gMin = Emulation.settings.TYPICAL_MOBILE_THROTTLING_METRICS.targetLatency - 10;
-      const areLatenciesAll3G = allRequestLatencies.every(val =>
-          val === undefined || val.latency > latency3gMin);
+      const areLatenciesAll3G = firstRequestLatencies.every(val => val.latency > latency3gMin);
```
Comment (paulirish) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: i missed this before but we're doing a `>` comparison on strings.

let's keep the `latency` value from above, separately from the toLocaleString() version..

Comment (patrickhulce) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: yeah it threw me for a loop at first too but type coercion does its thang haha

File: lighthouse-core/lib/url-shim.js (hunk 6 above)
Comment (patrickhulce) [inline lighthouse-core/lib/url-shim.js]: done

File: lighthouse-core/audits/load-fast-enough-for-pwa.js (hunk 7 above)
Comment (patrickhulce) [inline lighthouse-core/audits/load-fast-enough-for-pwa.js]: done

**Pull request merged**